# L5: Information extraction

Information extraction (IE) is the task of identifying named entities and semantic relations between these entities in text data. In this lab we will focus on two sub-tasks in IE, **named entity recognition** (identifying mentions of entities) and **entity linking** (matching these mentions to entities in a knowledge base).

**Reminder about our [Rules for hand-in assignments](https://www.ida.liu.se/~TDDE16/exam.en.shtml#handins) and the [Policy on cheating and plagiarism](https://www.ida.liu.se/~TDDE16/exam.en.shtml#cheating)**

We start by loading spaCy:

In [222]:
import spacy

nlp = spacy.load('en_core_web_sm')

The data that we will be using has been tokenized following the conventions of the [Penn Treebank](ftp://ftp.cis.upenn.edu/pub/treebank/public_html/tokenization.html), and we need to prevent spaCy from using its own tokenizer on top of this. We therefore override spaCy&rsquo;s tokenizer with one that simply splits on space.

In [223]:
from spacy.tokens import Doc

class WhitespaceTokenizer(object):
    def __init__(self, vocab):
        self.vocab = vocab

    def __call__(self, text):
        return Doc(self.vocab, words=text.split(' '))

nlp.tokenizer = WhitespaceTokenizer(nlp.vocab)

## Data set

The main data set for this lab is a collection of news wire articles in which mentions of named entities have been annotated with page names from the [English Wikipedia](https://en.wikipedia.org/wiki/). The next code cell loads the training and the development parts of the data into Pandas data frames.

In [224]:
import bz2
import csv
import pandas as pd

with bz2.open('ner-train.tsv.bz2', 'rt') as source:
    df_train = pd.read_csv(source, sep='\t', quoting=csv.QUOTE_NONE)

with bz2.open('ner-dev.tsv.bz2', 'rt') as source:
    df_dev = pd.read_csv(source, sep='\t', quoting=csv.QUOTE_NONE)

Each row in these two data frames corresponds to one mention of a named entity and has five columns:

1. a unique identifier for the sentence containing the entity mention
2. the pre-tokenized sentence, with tokens separated by spaces
3. the start position of the token span containing the entity mention
4. the end position of the token span (exclusive, as in Python list indexing)
5. the entity label; either a Wikipedia page name or the generic label `--NME--`

The following cell prints the first five samples from the training data:

In [226]:
df_train.head()

sentence_id                                          sentence  beg  end  \
0    0000-000  EU rejects German call to boycott British lamb .    0    1   
1    0000-000  EU rejects German call to boycott British lamb .    2    3   
2    0000-000  EU rejects German call to boycott British lamb .    6    7   
3    0000-001                                   Peter Blackburn    0    2   
4    0000-002                               BRUSSELS 1996-08-22    0    1   

            label  
0         --NME--  
1         Germany  
2  United_Kingdom  
3         --NME--  
4        Brussels

In this sample, we see that the first sentence is annotated with three entity mentions:

* the span 0–1 &lsquo;EU&rsquo; is annotated as a mention but only labelled with the generic `--NME--`
* the span 2–3 &lsquo;German&rsquo; is annotated with the page [Germany](http://en.wikipedia.org/wiki/Germany)
* the span 6–7 &lsquo;British&rsquo; is annotated with the page [United_Kingdom](http://en.wikipedia.org/wiki/United_Kingdom)

## Problem 1: Evaluation measures

To warm up, we ask you to write code to print the three measures that you will be using for evaluation:

In [227]:
def evaluation_report(gold, pred):
    """Print precision, recall, and F1 score.
    
    Args:
        gold: The set with the gold-standard values.
        pred: The set with the predicted values.
    
    Returns:
        Nothing, but prints the precision, recall, and F1 values computed
        based on the specified sets.
    """
    # TODO: Replace the next line with your own code
    TP = len(gold & pred) # intersection
    TN = 0
    FP = len(pred - gold)
    FN = len(gold - pred)

    precision = TP / (TP + FP)
    recall = TP / (TP + FN)
    F1_score = (2 * precision * recall) / (precision + recall)
    
    print("precission: ",precision, ", recall: ", recall, ", F1_score: ", F1_score )


To test your code, you can run the following cell:

In [229]:
evaluation_report(set(range(3)), set(range(5)))

precission:  0.6 , recall:  1.0 , F1_score:  0.7499999999999999


This should give you a precision of 60%, a recall of 100%, and an F1-value of 75%.

## Problem 2: Span recognition

One of the first tasks that an information extraction system has to solve is to locate and classify (mentions of) named entities, such as persons and organizations. Here we will tackle the simpler task of recognizing **spans** of tokens that contain an entity mention, without the actual entity label.

The English language model in spaCy features a full-fledged [named entity recognizer](https://spacy.io/usage/linguistic-features#named-entities) that identifies a variety of entities, and can be updated with new entity types by the user. Your task in this problem is to evaluate the performance of this component when predicting entity spans in the development data.

Start by implementing a generator function that yields the gold-standard spans in a given data frame.

**Hint:** The Pandas method [`itertuples()`](https://pandas.pydata.org/pandas-docs/version/0.17.0/generated/pandas.DataFrame.itertuples.html) is useful when iterating over the rows in a DataFrame.

In [276]:
def gold_spans(df):
    """Yield the gold-standard mention spans in a data frame.

    Args:
        df: A data frame.

    Yields:
        The gold-standard mention spans in the specified data frame as
        triples consisting of the sentence id, start position, and end
        position of each span.
    """
    # TODO: Replace the next line with your own code
    for row in df.itertuples(index = False):
        yield row.sentence_id, row.beg, row.end
    

To test your code, you can count the spans yielded by your function. When called on the development data, you should get a total of 5,917 unique triples. The first triple and the last triple should be

    ('0946-000', 2, 3)
    ('1161-010', 1, 3)  

In [280]:
spans_dev_gold = set(gold_spans(df_dev))
print(len(spans_dev_gold))


5917


----set() function will deduplicate and reorder the list.

----use list() function to get the original order.

In [281]:
list(gold_spans(df_dev))[0]


('0946-000', 2, 3)

In [282]:
list(gold_spans(df_dev))[-1]


('1161-010', 1, 3)

Your next task is to write code that calls spaCy to predict the named entities in the development data, and to evaluate the accuracy of these predictions in terms of precision, recall, and F1. Print these scores using the function that you wrote for Problem&nbsp;1.

In [287]:
# TODO: Write code here to run and evaluate the spaCy named entity recognizer on the development data
def spaCy_predict(df):
    for row in df.itertuples():
        doc = nlp(row.sentence)
        for ent in doc.ents:
            yield row.sentence_id, ent.start, ent.end

spaCy_predictions = set(spaCy_predict(df_dev))
evaluation_report(gold = spans_dev_gold, pred = spaCy_predictions)


precission:  0.5213954072029113 , recall:  0.702213959776914 , F1_score:  0.5984444764511019


## Problem 3: Error analysis

As you were able to see in Problem&nbsp;2, the span accuracy of the named entity recognizer is far from perfect. In particular, only slightly more than half of the predicted spans are correct according to the gold standard. Your next task is to analyse this result in more detail.

Here is a function that prints the false positives as well as the false negatives spans for a data frame, given a reference set of gold-standard spans and a candidate set of predicted spans.

In [288]:
from collections import defaultdict

def error_report(df, spans_gold, spans_pred):
    false_pos = defaultdict(list)
    for s, b, e in spans_pred - spans_gold:
        false_pos[s].append((b, e))
    false_neg = defaultdict(list)
    for s, b, e in spans_gold - spans_pred:
        false_neg[s].append((b, e))
    for row in df.drop_duplicates('sentence_id').itertuples():
        if row.sentence_id in false_pos or row.sentence_id in false_neg:
            print('Sentence:', row.sentence)
            for b, e in false_pos[row.sentence_id]:
                print('  FP:', ' '.join(row.sentence.split()[b:e]))
            for b, e in false_neg[row.sentence_id]:
                print('  FN:', ' '.join(row.sentence.split()[b:e]))

Use this function to inspect and analyse the errors that the automated prediction makes. Can you see any patterns? Base your analysis on the first 500 rows of the training data. Summarize your observations in a short text.

In [289]:
# TODO: Write code here to do your analysis

error_report(df_dev.iloc[:500,], set(gold_spans(df_dev)),set(spaCy_predict(df_dev)))


Sentence: CRICKET - LEICESTERSHIRE TAKE OVER AT TOP AFTER INNINGS VICTORY .
  FN: LEICESTERSHIRE
Sentence: LONDON 1996-08-30
  FP: 1996-08-30
Sentence: West Indian all-rounder Phil Simmons took four for 38 on Friday as Leicestershire beat Somerset by an innings and 39 runs in two days to take over at the head of the county championship .
  FP: Friday
  FP: 39
  FP: four
  FP: 38
  FP: two days
Sentence: After bowling Somerset out for 83 on the opening morning at Grace Road , Leicestershire extended their first innings by 94 runs before being bowled out for 296 with England discard Andy Caddick taking three for 83 .
  FP: the opening morning
  FP: 296
  FP: three
  FP: 83
  FP: 83
  FP: first
  FP: 94
Sentence: Trailing by 213 , Somerset got a solid start to their second innings before Simmons stepped in to bundle them out for 174 .
  FP: 213
  FP: second
  FP: 174
Sentence: Essex , however , look certain to regain their top spot after Nasser Hussain and Peter Such gave them a firm grip

*TODO: Write a short text that summarises the errors that you observed*

----The main reason for the error is the influence of numbers, such as: date, time, percentage, ordinal number, number, etc.


Now, use the insights from your error analysis to improve the automated prediction that you implemented in Problem&nbsp;2. While the best way to do this would be to [update spaCy&rsquo;s NER model](https://spacy.io/usage/linguistic-features#updating) using domain-specific training data, for this lab it suffices to write code to post-process the output produced by spaCy. To filter out specific labels it is useful to know the named entity label scheme, which can be found in the [model's documentation](https://spacy.io/models/en#en_core_web_sm). You should be able to improve the F1 score from Problem&nbsp;2 by at last 15 percentage points.

--original types:

---['CARDINAL', 'DATE', 'EVENT', 'FAC', 'GPE', 'LANGUAGE', 'LAW', 'LOC', 'MONEY', 'NORP', 'ORDINAL', 'ORG', 'PERCENT', 'PERSON', 'PRODUCT', 'QUANTITY', 'TIME', 'WORK_OF_ART']

--number related types:

---['TIME', 'DATE', 'MONEY', 'CARDINAL', 'QUANTITY', 'ORDINAL']

--we remove these number related types.


In [304]:
# TODO: Write code here to run and evaluate the spaCy named entity recognizer on the development data
def improved_spaCy_predict(df):
    for row in df.itertuples():
        doc = nlp(row.sentence)
        for ent in doc.ents:
            # here we remove these number related types.
            # ['TIME', 'DATE', 'MONEY', 'CARDINAL', 'QUANTITY', 'ORDINAL', 'PERCENT']
            if ent.label_ not in ['TIME', 'DATE', 'MONEY', 'CARDINAL', 'QUANTITY', 'ORDINAL', 'PERCENT']:
                yield row.sentence_id, ent.start, ent.end

improved_spaCy_predictions = set(improved_spaCy_predict(df_dev))
evaluation_report(gold = spans_dev_gold, pred = improved_spaCy_predictions)


precission:  0.847284605961617 , recall:  0.701368936961298 , F1_score:  0.7674526121128065


In [305]:
0.7674526121128065 - 0.5984444764511019

0.16900813566170458

----As we can see, after removing these number related types, we improved approximately 16.9% F1_score.

Show that you achieve the performance goal by reporting the evaluation measures that you implemented in Problem&nbsp;1.

Before going on, we ask you to store the outputs of the improved named entity recognizer on the development data in a new data frame. This new frame should have the same layout as the original data frame for the development data that you loaded above, but should contain the *predicted* start and end positions for each token span, rather than the gold positions. As the `label` of each span, you can use the special value `--NME--`.

In [306]:
df_dev.head()

sentence_id                                           sentence  beg  end  \
0    0946-000  CRICKET - LEICESTERSHIRE TAKE OVER AT TOP AFTE...    2    3   
1    0946-001                                  LONDON 1996-08-30    0    1   
2    0946-002  West Indian all-rounder Phil Simmons took four...    0    2   
3    0946-002  West Indian all-rounder Phil Simmons took four...    3    5   
4    0946-002  West Indian all-rounder Phil Simmons took four...   12   13   

                                label  
0  Leicestershire_County_Cricket_Club  
1                              London  
2            West_Indies_cricket_team  
3                        Phil_Simmons  
4  Leicestershire_County_Cricket_Club

In [313]:
# TODO: Write code here to store the predicted spans in a new data frame

# Create an empty DataFrame.
# Just remove the old records, but keep the column names.
new_df_dev = df_dev.drop(index=df_dev.index)

#add new row to the end of the empty DataFrame.
for id, beg, end in improved_spaCy_predictions:
    new_df_dev.loc[len(new_df_dev.index)] = [id, df_dev[df_dev['sentence_id'] == id].iloc[0]['sentence'], beg, end, '--NME--']

new_df_dev


sentence_id                                           sentence  beg  end  \
0       0966-092                  3. Fatima Yusuf ( Nigeria ) 51.43    1    3   
1       1094-030                             NEW YORK AT CALIFORNIA    0    2   
2       1157-000        Nato declines comment on fighting in Iraq .    6    7   
3       0966-130             6= Nele Zilinskiene ( Lithuania ) 1.91    4    5   
4       1076-029                          Obilic 2 Zeleznik ( B ) 0    0    3   
...          ...                                                ...  ...  ...   
4893    1064-007  With the first day of the league season briefl...   45   46   
4894    0973-002  Australia scored 228 for nine wickets in their...   11   13   
4895    0946-012  Australian Tom Moody took six for 82 but Chris...   20   21   
4896    1042-002  Police in southwest China have arrested 30 peo...    3    4   
4897    0972-013         J. Gillespie st Kaluwitharana b Chandana 6    0    4   

        label  
0     --NME--  
1     --NME--  
2     --NME--  
3     --NME--  
4     --NME--  
...       ...  
4893  --NME--  
4894  --NME--  
4895  --NME--  
4896  --NME--  
4897  --NME--  

[4898 rows x 5 columns]

## Problem 4: Entity linking

Now that we have a method for predicting mention spans, we turn to the task of **entity linking**, which amounts to predicting the knowledge base entity that is referenced by a given mention. In our case, for each span we want to predict the Wikipedia page that this mention references.

Start by extending the generator function that you implemented in Problem&nbsp;2 to labelled spans.

In [314]:
def gold_mentions(df):
    """Yield the gold-standard mentions in a data frame.

    Args:
        df: A data frame.

    Yields:
        The gold-standard mention spans in the specified data frame as
        quadruples consisting of the sentence id, start position, end
        position and entity label of each span.
    """
    # TODO: Replace the next line with your own code
    # --just add one more label column
    for row in df.itertuples():
        yield row.sentence_id, row.beg, row.end, row.label



A naive baseline for entity linking on our data set is to link each mention span to the Wikipedia page name that we get when we join the tokens in the span by underscores, as is standard in Wikipedia page names. Suppose, for example, that a span contains the two tokens

    Jimi Hendrix

The baseline Wikipedia page name for this span would be

    Jimi_Hendrix

Implement this naive baseline and evaluate its performance. Print the evaluation measures that you implemented in Problem&nbsp;1.

**Here and in the remainder of this lab, you should base your entity predictions on the predicted spans that you computed in Problem&nbsp;3.**

In [317]:
# TODO: Write code here to implement the baseline
def naive_baseline_predict(df):
    for row in df.itertuples():
        label = "_".join(row.sentence.split(' ')[row.beg : row.end])
        yield row.sentence_id, row.beg, row.end, label


In [318]:
gold_standard_mentions = set(gold_mentions(df_dev))
naive_baseline_prediction = set(naive_baseline_predict(new_df_dev))

evaluation_report(gold = gold_standard_mentions, pred = naive_baseline_prediction)


precission:  0.31400571661902815 , recall:  0.25992901808348823 , F1_score:  0.2844197873324087


## Problem 5: Extending the training data using the knowledge base

State-of-the-art approaches to entity linking exploit information in knowledge bases. In our case, where Wikipedia is the knowledge base, one particularly useful type of information are links to other Wikipedia pages. In particular, we can interpret the anchor texts (the highlighted texts that you click on) as mentions of the entities (pages) that they link to. This allows us to harvest long lists of mention–entity pairings.

The following cell loads a data frame summarizing anchor texts and page references harvested from the first paragraphs of the English Wikipedia. The data frame also contains all entity mentions in the training data (but not the development or the test data).

In [319]:
with bz2.open('kb.tsv.bz2', 'rt') as source:
    df_kb = pd.read_csv(source, sep='\t', quoting=csv.QUOTE_NONE)

To understand what information is available in this data, the following cell shows the entry for the anchor text `Sweden`.

In [320]:
df_kb.loc[df_kb.mention == 'Sweden'].iloc[0]


mention      Sweden
entity       Sweden
prob       0.985768
Name: 17436, dtype: object

In [322]:
df_kb.head()


mention                   entity  prob
0      000 Guineas      2000_Guineas_Stakes   1.0
1            10 00     United_States_dollar   1.0
2      126 million     United_States_dollar   1.0
3     13th dynasty  Middle_Kingdom_of_Egypt   1.0
4  14th Dalai Lama          14th_Dalai_Lama   1.0

As you can see, each row of the data frame contains a pair $(m, e)$ of a mention $m$ and an entity $e$, as well as the conditional probability $P(e|m)$ for mention $m$ referring to entity $e$. These probabilities were estimated based on the frequencies of mention–entity pairs in the knowledge base. The example shows that the anchor text &lsquo;Sweden&rsquo; is most often used to refer to the entity [Sweden](http://en.wikipedia.org/wiki/Sweden), but in a few cases also to refer to Sweden&rsquo;s national football and ice hockey teams. Note that references are sorted in decreasing order of probability, so that the most probable pairing come first.

Implement an entity linking method that resolves each mention to the most probable entity in the data frame. If the mention is not included in the data frame, you can predict the generic label `--NME--`. Print the precision, recall, and F1 of your method using the function that you implemented for Problem&nbsp;1.

In [221]:
def gold_mentions(df):
    for row in df.itertuples():
        yield row.mention, row.entity
gold_standard_mentions = set(gold_mentions(df_kb))
print(len(gold_standard_mentions))
gold_standard_mentions

23359


{('Spanish Chavacano', 'Spanish_language'),
 ('Soviet regime', 'Soviet_Union'),
 ('WSC World XI', 'World_Series_Cricket_World_XI'),
 ('Vörös Lobogó', 'MTK_Budapest_FC'),
 ('Sale Victoria', 'Victoria_(Australia)'),
 ('Unita', 'UNITA'),
 ('John Newcombe', 'John_Newcombe'),
 ('Republic of France', 'France'),
 ('Todd Woodbridge', 'Todd_Woodbridge'),
 ('Oaxaca region', 'Oaxaca'),
 ('Firenze', 'Florence'),
 ('Spanish', 'Spaniards'),
 ('Venezuela', 'Venezuela'),
 ('Berlin 1936', '1936_Summer_Olympics'),
 ('Marseillais', 'Marseille'),
 ('Genoa', 'Genoa_C.F.C.'),
 ('modern country', 'France'),
 ('Margaret Crowley', 'Margaret_Crowley_(athlete)'),
 ('Tauziat', 'Nathalie_Tauziat'),
 ('CNN Travel', 'CNN'),
 ('Rugby League', 'Rugby_league'),
 ('Brenden', 'Brenden_Pappas'),
 ('Naziism', 'Nazi_Party'),
 ('Soviet side', 'Soviet_Union'),
 ('major league level', 'Major_League_Baseball'),
 ('Ericsson AB', 'Ericsson'),
 ('Bordèu', 'Bordeaux'),
 ('1992 Barcelona Games', '1992_Summer_Olympics'),
 ('GOP', 'Re

In [200]:
# TODO: Write code here to implement the "most probable entity" method.
def most_probable_entity(df):
    for row in df.itertuples():
        df[df.mention == 'Sweden'].iloc[0]

        
most_probable_entity(df_kb)

SyntaxError: incomplete input (1384604170.py, line 2)

## Problem 6: Context-sensitive disambiguation

Consider the entity mention &lsquo;Lincoln&rsquo;. The most probable entity for this mention turns out to be [Lincoln, Nebraska](http://en.wikipedia.org/Lincoln,_Nebraska); but in pages about American history, we would be better off to predict [Abraham Lincoln](http://en.wikipedia.org/Abraham_Lincoln). This suggests that we should try to disambiguate between different entity references based on the textual context on the page from which the mention was taken. Your task in this last problem is to implement this idea.

Set up a dictionary that contains, for each mention $m$ that can refer to more than one entity $e$, a separate Naive Bayes classifier that is trained to predict the correct entity $e$, given the textual context of the mention. As the prior probabilities of the classifier, choose the probabilities $P(e|m)$ that you used in Problem&nbsp;5. To let you estimate the context-specific probabilities, we have compiled a data set with mention contexts:

In [ ]:
with bz2.open('contexts.tsv.bz2') as source:
    df_contexts = pd.read_csv(source, sep='\t', quoting=csv.QUOTE_NONE)

This data frame contains, for each ambiguous mention $m$ and each knowledge base entity $e$ to which this mention can refer, up to 100 randomly selected contexts in which $m$ is used to refer to $e$. For this data, a **context** is defined as the 5 tokens to the left and the 5 tokens to the right of the mention. Here are a few examples:

In [ ]:
df_contexts.head()

Note that, in each context, the position of the mention is indicated by the `@` symbol.

From this data frame, it is easy to select the data that you need to train the classifiers – the contexts and corresponding entities for all mentions. To illustrate this, the following cell shows how to select all contexts that belong to the mention &lsquo;Lincoln&rsquo;:

In [ ]:
df_contexts.context[df_contexts.mention == 'Lincoln']

Implement the context-sensitive disambiguation method and evaluate its performance. Here are some more hints that may help you along the way:

**Hint 1:** The prior probabilities for a Naive Bayes classifier can be specified using the `class_prior` option. You will have to provide the probabilities in the same order as the alphabetically sorted class (entity) names.

**Hint 2:** Not all mentions in the knowledge base are ambiguous, and therefore not all mentions have context data. If a mention has only one possible entity, pick that one. If a mention has no entity at all, predict the `--NME--` label.

In [ ]:
# TODO: Write code here to implement the context-sensitive disambiguation method

You should expect to see a small (around 1&nbsp;unit) increase in both precision, recall, and F1.

## Reflection questions

The following reflection questions will help you prepare for the diagnostic test. Answer each of them in the form of a short text and put your answers in the cell below. You will get feedback on your answers from your lab assistant.

**RQ 5.1:** In Problem&nbsp;3, you did an error analysis on the task of recognizing text spans mentioning named entities. Summarize your results. Pick one type of error that you observed. How could you improve the model&rsquo;s performance on this type of error? What resources (such as domain knowledge, data, compute) would you need to implement this improvement?

**RQ 5.2:** Thinking back about Problem&nbsp;6, explain what the word *context* refers to in the task addressed there, and how context can help to disambiguate between different entities. Suggest other types of context that you could use for disambiguation.

**RQ 5.3:** One type of entity mentions that we did not cover explicitly in this lab are pronouns. As an example, consider the sentence pair *Ruth Bader Ginsburg was an American jurist*. *She served as an associate justice of the Supreme Court from 1993 until her death in 2020*. What facts would you want to extract from this sentence pair? How do pronouns make fact extraction hard?

*TODO: Enter your answers here*

**This was the last lab in the Text Mining course. Congratulations! 🥳**